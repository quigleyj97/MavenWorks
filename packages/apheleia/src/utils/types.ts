type DataType = 'layout-part' | 'stack-panel';

interface LayoutPartData {
    type: 'layout-part';
    title: string;
    showTitlebar: boolean;
}

interface StackPanelData {
    type: 'stack-panel';
    children: BaseRegionData[];
}

type BaseRegionData = LayoutPartData | StackPanelData;
