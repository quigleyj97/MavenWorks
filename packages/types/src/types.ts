import { GenericType } from "./generic";
import { Type } from "./type";

/**
 * Default built-in type definitions.
 *
 * @export
 * @class Types
 */
export class Types {
    /** The top type, aka "Any".
     *
     * This may represent any value, and any value may be assigned to it.
     *
     * @static
     * @memberof Types
     */
    public static Unknown = Type.CreateType("Unknown");

    /** A generic number type.
     *
     * @static
     * @memberof Types
     */
    public static Number = Type.CreateType("Number", Types.Unknown);
    /** A type for an integer.
     *
     * In JS, values of this type should be BigInts.
     *
     * @static
     * @memberof Types
     */
    public static Integer = Type.CreateType("Integer", Types.Number);
    /** A double-precision floating-point type.
     *
     * @static
     * @memberof Types
     */
    public static Float = Type.CreateType("Float", Types.Number);
    /**
     * A decimal type.
     *
     * This is useful for avoiding floating-point errors in decimal math, eg
     * financial pricing data. In JS, values of this type are Decimal.js
     * instances. cf. https://github.com/MikeMcl/decimal.js/
     *
     * @static
     * @memberof Types
     */
    public static Decimal = Type.CreateType("Decimal", Types.Number);
    /**
     * A string type.
     *
     * @static
     * @memberof Types
     */
    public static String = Type.CreateType("String", Types.Unknown);

    /**
     * An Object type, representing values that are not considered 'primitives'.
     *
     * @static
     * @memberof Types
     */
    public static Object = Type.CreateType("Object", Types.Unknown);

    /**
     * A Date type, representing a calendar date with no associated time.
     *
     * @static
     * @memberof Types
     */
    public static Date = Type.CreateType("Date", Types.Object);

    /**
     * A DateTime type, representing a combined calendar date and time-of-day.
     *
     * @static
     * @memberof Types
     */
    public static DateTime = Type.CreateType("DateTime", Types.Object);

    /**
     * A Time type, representing a time-of-day (but not a calendar date).
     *
     * @static
     * @memberof Types
     */
    public static Time = Type.CreateType("Time", Types.Object);

    /**
     * A generic factory to create Array types.
     *
     * @static
     * @memberof Types
     */
    public static Array = GenericType.CreateGenericFactory("Array", Types.Object, 1);

    /**
     * A generic factory to create Map types.
     *
     * @static
     * @memberof Types
     */
    public static Map = GenericType.CreateGenericFactory("Map", Types.Object, 2);
}