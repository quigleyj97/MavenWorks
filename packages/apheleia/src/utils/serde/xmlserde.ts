import { BaseRegionData, LayoutData, LayoutPartData, StackPanelData } from "../types";
import { ILayoutSerde } from "./interface";

export class XmlLayoutSerde implements ILayoutSerde {
    private static readonly VERSION_MAJOR = 0;
    private static readonly VERSION_MINOR = 1;
    private static readonly VERSION_NUMBER = XmlLayoutSerde.VERSION_MAJOR +
        "." + XmlLayoutSerde.VERSION_MINOR;
    private static readonly LAYOUT_TAG = "LAYOUT";
    private static readonly STACKPANEL_TAG = "STACK";
    private static readonly PART_TAG = "PART";

    public loadFromString(data: string): LayoutData {
        const reader = new DOMParser();
        let doc: Document;
        try {
            doc = reader.parseFromString(data, "application/xml");
        }
        catch (err) {
            throw Error("Faled to parse XML: " + err); // nothing better to do
        }
        const errorDoc = doc.getElementsByTagName("parsererror");
        if (errorDoc.length > 0) {
            // There's an error in the document, but the user agent was either
            // able to partially parse it or opted not to throw an error.
            throw new Error("ParserError: " + errorDoc);
        }
        // Return the root element.
        const layout = this.walkXmlTree(doc);
        return {
            root: layout
        };
    }

    public dumpToString(layout: LayoutData): string {
        const doc = new Document();
        const root = doc.createElement(XmlLayoutSerde.LAYOUT_TAG);
        root.setAttribute("version", XmlLayoutSerde.VERSION_NUMBER);

        const rootStack = this.makeXmlNode(doc, layout.root);
        root.appendChild(rootStack);
        doc.appendChild(root);

        const serializer = new XMLSerializer();
        return serializer.serializeToString(doc);
    }

    private makeXmlNode(doc: XMLDocument, region: BaseRegionData) {
        let node: Element;
        switch (region.type) {
            case 'layout-part':
                node = this.makePartXmlNode(doc, region);
                break;
            case 'stack-panel':
                node = this.makeStackXmlNode(doc, region);
                break;
        }
        return node;
    }

    private makePartXmlNode(doc: XMLDocument, region: LayoutPartData): Element {
        const node = doc.createElement(XmlLayoutSerde.PART_TAG);
        node.setAttribute("title", region.title);
        node.setAttribute("id", region.id);
        node.setAttribute("showTitlebar", ""+region.showTitlebar);
        return node;
    }

    private makeStackXmlNode(doc: XMLDocument, region: StackPanelData): Element {
        const node = doc.createElement(XmlLayoutSerde.STACKPANEL_TAG);
        node.setAttribute("direction", region.direction);
        for (let i = 0; i < region.children.length; i++) {
            const childRegion = region.children[i];
            const childNode = this.makeXmlNode(doc, childRegion);
            childNode.setAttribute("stackSize", region.stackSizes[i]);
            node.appendChild(childNode);
        }
        return node;
    }

    private walkXmlTree(xmldoc: XMLDocument) {
        const root = xmldoc.children[0];
        if (root == null || root.tagName.toLocaleUpperCase() != XmlLayoutSerde.LAYOUT_TAG) {
            throw new Error("Invalid root element: " + root);
        }
        const version = root.getAttribute("version");
        this.checkVersionString(version);
        return this.visitXmlNode(root.children[0]);
    }

    private checkVersionString(version: string | null) {
        if (version == null) {
            console.warn("Version missing, proceeding with parse anyway");
            return;
        }
        // since it's declared, check it
        const [major, minor] = version.split(".");
        if (!Number.isInteger(+major) || !Number.isInteger(+minor)) {
            throw Error("Invalid version string: " + version)
        }
        if (
            +major > XmlLayoutSerde.VERSION_MAJOR ||
            +minor > XmlLayoutSerde.VERSION_MINOR
        ) {
            throw Error(
                "Layout version too new: " + version + " is greater than " +
                XmlLayoutSerde.VERSION_NUMBER
            );
        }
    }

    private visitXmlNode(node: Element): BaseRegionData;
    private visitXmlNode(node: Element, parentRegion: StackPanelData): void;
    private visitXmlNode(node: Element, parentRegion?: StackPanelData) {
        const tagName = node.tagName.toUpperCase();
        let region: BaseRegionData;
        switch (tagName) {
            case XmlLayoutSerde.STACKPANEL_TAG:
                region = this.visitStackElement(node);
                break;
            case XmlLayoutSerde.PART_TAG:
                region = this.visitPartElement(node);
                break;
            default:
                region = this.visitUnknownElement(node);
                break;
        }
        if (parentRegion) {
            parentRegion.children.push(region);
            parentRegion.stackSizes.push(node.getAttribute("stackSize") ?? "1");
            return;
        } else {
            return region;
        }
    }

    private visitUnknownElement(node: Element): BaseRegionData {
        return {
            type: 'layout-part',
            id: '###error-part',
            title: 'Error: Unknown layout region: ' + node.tagName,
            showTitlebar: true
        };
    }

    private visitStackElement(node: Element): BaseRegionData {
        let direction = node.getAttribute("direction");
        if (direction != 'horizontal' && direction != 'vertical' && direction != null) {
            console.warn("Invalid direction for stack panel: " + direction);
            console.warn(node);
            direction = null;
        }
        const region: StackPanelData = {
            type: 'stack-panel',
            children: [],
            direction: direction ?? 'horizontal',
            stackSizes: []
        };
        for (const child of node.children) {
            this.visitXmlNode(child, region);
        }
        return region;
    }

    private visitPartElement(node: Element): BaseRegionData {
        let id = node.getAttribute("id") ?? "###error-part";
        let title = node.getAttribute("title") ?? "";
        let showTitlebar = this.stringToBool(
            node.getAttribute("showTitlebar") ?? "true"
        );

        return {
            type: 'layout-part',
            id,
            title,
            showTitlebar
        };
    }

    private stringToBool(str: string): boolean {
        switch (str.toLowerCase()) {
            case "t":
            case "true":
            case "1":
                return true;
            case "f":
            case "false":
            case "0":
                return false;
        }
        console.warn("Cannot easily map " + str + " to a boolean value, assuming falsy");
        return false;
    }
}