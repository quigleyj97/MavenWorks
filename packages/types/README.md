# Types

Types, type checking, and type comparison mechanisms
