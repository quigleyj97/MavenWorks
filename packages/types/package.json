{
  "name": "@mavenomics/types",
  "version": "0.1.2",
  "description": "Types and Type Serialization",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "repository": "https://github.com/Mavenomics/MavenWorks",
  "author": "Mavenomics, Inc",
  "license": "GPL-3.0",
  "files": [
    "lib/"
  ],
  "scripts": {
    "build": "tsc --build",
    "build:ci": "echo pass",
    "clean": "rimraf ./lib *.tsbuildinfo",
    "lint": "tslint --project .",
    "test": "jest",
    "docs": "typedoc --tsconfig ./docs.tsconfig.json"
  },
  "devDependencies": {},
  "dependencies": {}
}
