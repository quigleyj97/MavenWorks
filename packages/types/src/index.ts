export { Types } from "./types";
export { Type } from "./type";
export { GenericType } from "./generic";
