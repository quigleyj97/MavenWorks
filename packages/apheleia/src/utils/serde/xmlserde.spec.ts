import * as fs from "fs";
import { LayoutData } from "../types";
import { XmlLayoutSerde } from "./xmlserde"

const TEST_FILE = fs.readFileSync("./testdata/layout.xml", "utf8");

const TEST_LAYOUT: LayoutData = {
    root: {
        type: 'stack-panel',
        children: [{
            type: 'layout-part',
            showTitlebar: false,
            title: 'test-part',
            id: 'test-part'
        }],
        stackSizes: [
            "200px"
        ],
        direction: 'vertical'
    }
};

describe("XmlLayoutSerde", () => {
    const serde = new XmlLayoutSerde();

    test("rejects too-new-version", () => {
        const TEST_DOC = `<Layout version="1000.1000"></Layout>`;
        expect(() => serde.loadFromString(TEST_DOC)).toThrowError("Layout version too new");
    });

    test("parses doc", () => {
        expect(serde.loadFromString(TEST_FILE)).toMatchSnapshot();
    });

    test("serializes layout", () => {
        expect(serde.dumpToString(TEST_LAYOUT)).toMatchSnapshot();
    })
})