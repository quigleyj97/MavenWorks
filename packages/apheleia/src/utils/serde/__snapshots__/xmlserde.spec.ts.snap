// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XmlLayoutSerde parses doc 1`] = `
Object {
  "root": Object {
    "children": Array [
      Object {
        "id": "input_1",
        "showTitlebar": true,
        "title": "Input",
        "type": "layout-part",
      },
      Object {
        "id": "output_grid",
        "showTitlebar": false,
        "title": "SlickGrid",
        "type": "layout-part",
      },
      Object {
        "children": Array [
          Object {
            "id": "slider_range",
            "showTitlebar": true,
            "title": "Range",
            "type": "layout-part",
          },
          Object {
            "id": "spinner",
            "showTitlebar": true,
            "title": "Category",
            "type": "layout-part",
          },
        ],
        "direction": "vertical",
        "stackSizes": Array [
          "1.5em",
          "1.5em",
        ],
        "type": "stack-panel",
      },
    ],
    "direction": "horizontal",
    "stackSizes": Array [
      "1",
      "1",
      "1",
    ],
    "type": "stack-panel",
  },
}
`;

exports[`XmlLayoutSerde serializes layout 1`] = `"<LAYOUT version=\\"0.1\\"><STACK direction=\\"vertical\\"><PART title=\\"test-part\\" id=\\"test-part\\" showTitlebar=\\"false\\" stackSize=\\"200px\\"/></STACK></LAYOUT>"`;
